### DISCUSSION TITLE
Run TestNG xml suites in debug mode

### DISCUSSION THREAD

**Discussion by anjerukare** [mfussenegger/nvim-jdtls#313] [Q&A] [answered]
Hello, I'm looking for a way to run TestNG xml suites in debug mode.

Now I know how to just run TestNG xml suites by specifying jar package:
```
java -cp [path-to-jar] org.testng.TestNG [path-to-xml-suite]
```
I also found parameter that allows to debug it remotely: `-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=1044`.

I guess next step is attaching. How can I attach to debug session from Neovim?

**Comment by mfussenegger** [comment, marked as answer]
You need to create an attach configuration:

```lua
local dap = require('dap')
dap.configurations.java = {
  {
    type = 'java';
    request = 'attach';
    name = "Debug (Attach) - Remote";
    hostName = "127.0.0.1";
    port = 1044;
  },
}
```

The `port` must match whatever you're using in `address=1044`
See https://github.com/mfussenegger/nvim-dap/wiki/Java#configuration

If you're using a multi-module project you'll also have to add a `projectName = '<yourproject>'` property.

You might want to change `suspend=n` to `suspend=y` to ensure it waits for a debugger to attach before running.


If you follow the vscode-java-test setup instructions (https://github.com/mfussenegger/nvim-jdtls#vscode-java-test-installation) you should also be able to use: `require'jdtls'.test_class()` and `require'jdtls'.test_nearest_method()` to trigger tests directly from within neovim
